### DISCUSSION TITLE
Example C# Project

### DISCUSSION THREAD

**Discussion by BigMack-1** [uhppoted/uhppoted#42] [General] [answered]
Hi There,
Just wondering if anyone has a Visual Studio C# example that implements the DLL?

Something that connects to the Board and updates a PIN and\or receives an entry\exit access.

Cheers

**Comment by uhppoted** [comment]
Hi,

Not to the best of my knowledge - I'm on a Mac so mostly just use _mono_ and command line builds for C#.

I could look into using _Visual Studio for Mac_ to put together a command line app if that would help at all?

**Reply by BigMack-1** [reply]
Thanks for the quick response. If you could provide a basic command line or a C# example that would be great.

I did notice a some files in one of C# testing directory’s of the repo that might demo the functions - I’ll check in more detail when I have a chance.

On another note - do you know if it’s possible to change the status of a user from the API? Ie the board registers them as IN - could you alter this to OUT or vice versa.

**Comment by uhppoted** [comment]
Hi,

There is a C# command line application in the [examples](https://github.com/uhppoted/uhppoted-dll/blob/main/examples/csharp/examples.cs) - it isn't _Visual Studio_ project but does have an example for every API function. I'll see if I can get a _Visual Studio_ project up and running though (am a bit swamped at the moment I'm afraid, so it will probably only happen either later this week or early next week).

re. IN/OUT. Not as far as I'm aware - have just been through the SDK with a fine-tooth comb and there doesn't seem to be any function in the public API that gets/sets a users location. There _may_ be a manufacturer-only function but TBH I would expect the in/out state of a user to be managed by the application because most installations have multiple controllers and e.g. a user could swipe into an area via one door and then swipe out via a door managed by a different controller.

**Comment by uhppoted** [comment]
Hi,

Just a quick note - I started looking into this and it seems like on Visual Studio for Windows can't import the DLL compiled by Go. No idea why (possibly something to do with the _gcc_ compiler) - I'll have to do some more digging to figure out why and how to fix it.

Tracking issue: [Example C# Project: Visual Studio for Windows cannot load DLL](https://github.com/uhppoted/uhppoted-dll/issues/9)

**Comment by BigMack-1** [comment]
Thanks for looking into this :)

Just for some background - I have another system that can grant users access (IN\OUT) but need PIN codes as a backup. I am looking to do something relatively straight forward as below:
  - Implement this as windows program.
  - Connect a web socket to another program - which sends and receives to the external system.
  - Add new users to the access board (they will have expiry times).
  - Receive IN\OUT updates locally from the access board
  - Alter a users state (or trigger an IN\OUT from the API) on the access board based on the external system.

It doesn't have to be in C# - but I figured its just a good way to package something up for a Windows machine.

Cheers

**Comment by uhppoted** [comment]
Hi,

Ok, so the good news is that I've managed to get a (very extremely) basic _C#_ console app working under both _Visual Studio for Mac_ and _Visual Studio for Windows_ - the project files are in the [_examples/csharp-vs_](https://github.com/uhppoted/uhppoted-dll/tree/main/examples/csharp-vs) folder.

Be warned though - at this stage it is nothing more than some **very** scrappy _proof-of-concept_ code - I'll put some time into polishing it up a bit and adding some functionality over the next week or two but hopefully the underlying stuff shouldn't change.

The critical bits turned out to be:

1. On Windows you **have** to have the _mingw_ compiler installed to compile the DLL. I've included the [DLL](https://github.com/uhppoted/uhppoted-dll/blob/main/examples/csharp-vs/vs-win/uhppoted.dll) in the examples but you may want to recompile it on your machine

2. For the launch configuration you need to:
   - Specify the working directory
   - Add the directory containing the DLL to the PATH environment variable (e.g. I put the DLL in the project directory and the set `PATH = %PATH%;${ProjectDir}`)

Other than that it's straightforward - if you wanted to start a solution from scratch all you would need to do is:
- Copy the [uhppoted.cs](https://github.com/uhppoted/uhppoted-dll/blob/main/bindings/csharp/uhppoted.cs) and [lookup.cs](https://github.com/uhppoted/uhppoted-dll/blob/main/bindings/csharp/lookup.cs) bindings files into your project
- Rebuild or copy the DLL into the project (or at least a known folder)
- Set the PATH environment variable to include the DLL folder 

In terms of what you want to do it mostly looks pretty doable - the only thing you might run into is that the board doesn't store the IN/OUT state of users so that's something you're going to have to manage yourself. Should be fun .. let me know how you go :-).

**Comment by twystd** [comment, marked as answer]
Hi,

Just let you know that there are now complete working examples for a _Visual Studio_ (both Mac and Windows) C# console app under the [_examples/csharp_](https://github.com/uhppoted/uhppoted-dll/tree/main/examples/csharp) folder.

The examples don't include the DLL because that more or less has to be built locally to link in the system specific libraries, but it's straightforward - it does require _Go_ to be installed and under _Windows_ **absolutely requires** the _mingw_ compiler:
```
git clone https://github.com/uhppoted/uhppoted-dll
cd uhppoted-dll
make build
```

Copy the _uhppoted.dll_ file from the _lib_ folder to your project folder and you should be good to go.

**Reply by threeeye** [reply]
Hi there,
Just a quick question (or clarification).
I installed mingw, when I run it, a cmd window opens, and I get:
'make' is not recognized as an internal or external command,
operable program or batch file.

I'm sure I'm doing something wrong...
Thanks in advance

**Reply by uhppoted** [reply]
Hi,

You probably just need to install _make_ as a separate thing - it's been a while but I seem to remember that it's not installed as part of _mingw_. If you're using _chocolatey_ then it's just:
```
choco install make
```

The _Makefile_ is simple enough that you can just run the commands manually if you prefer, but _make_ does simplify it a whole lot.

**Reply by threeeye** [reply]
Thanks for the reply.
I managed to 'make' the DLL, but when trying to add it to my .NET8 project I get an error: The reference is invalid or unsupported.
And adding it to a .NET 4.8 it crashes
Any suggestions?

Thanks again

EDIT:
I copied the DLL and the make file to my project, I think it works now, but GetDevices gives me 3 devices and non of them actually exists (I have about 50 controllers)

**Comment by uhppoted** [comment]
Hi,

Great! I've got a _TODO_ item to look into building the DLL on github but it will be a while before a I get to it so glad you got it working. I've only ever built this on .NET7 (_mono_) and .NET8 (_VS Windows/Mac_) - the C# foreign function interface has probably changed significantly since .NET 4.8 so not too surprised it doesn't work.

re. _GetDevices_. Hmm, that's weird - are they by any chance the test/simulation controllers (405419896, 303986753 and 201020304 )? 

However, the most likely reason is that `GetDevices` uses UDP broadcast so it can only find controllers on the local LAN. If the controllers are on a separate network segment they're not reachable by UDP broadcast and you need to add a list of controllers and their addresses when instantiating the `Uhppoted` instance - the [_mono_](https://github.com/uhppoted/uhppoted-dll/blob/main/examples/csharp/mono/examples.cs#L159-L178) example shows how to do that:
```
   ...
    static Controller[] controllers = { new Controller(405419896, "192.168.1.100"),
                                        new Controller(303986753, "192.168.1.100") };

   ...
   Uhppoted u = new Uhppoted("0.0.0.0", "255.255.255.255", "0.0.0.0:60001", 2500, controllers, true);
   ...
```

There are a couple of other things you can try:
- try increasing the timeout from 2500ms to say 5000ms and enable debugging when you instantiate the `Uhppoted` instance:
```
  ...
   Uhppoted u = new Uhppoted("0.0.0.0", "255.255.255.255", "0.0.0.0:60001", 5000, controllers, true);
   ...
```
- download the [CLI](https://github.com/uhppoted/uhppote-cli/releases) and run the _get-devices_ command:
```
uhppote-cli --debug get-devices
```

**Reply by threeeye** [reply]
Yes, I'm getting the test/simulation controllers.
I imported the uhppoted.dll and uhppoted-dll-cli.dll but I don't see where the test/simulation is configured

**Reply by uhppoted** [reply]
Well, that was totally a wild guess :-).

Ok (and assuming you are not actually running the _simulator_) - if you did a `make build` it will have built three versions of the DLL in the _lib_ folder:
- **_lib\uhppoted.dll_**
- _lib\debug\uhppoted.dll_
- _lib\tests\uhppoted.dll_

The one you want is _lib\uhppoted.dll_ and you don't need the _uhppoted-dll-cli.dll_. At a guess you are using either the _debug_ or _tests_ DLLs, both of which return a canned response:
- _lib\debug\uhppoted.dll_ is for use in debugging the weird ways languages can pack data when they invoke a DLL function
- _lib\tests\uhppoted.dll_  is for integrations tests and is used to validate the function calls are being invoked correctly.

**Reply by threeeye** [reply]
Thanks for the info.
But I still can't add the DLL as a reference in my project (VB.NET on .NET8 Core)
I get an error message: The reference is invalid or unsupported.

**Reply by uhppoted** [reply]
How did you resolve the "reference is invalid or unsupported" error before? 

It normally means you need to add it to the launch settings and possibly also the PATH environment variable, e.g.:
```
launchSettings.json
{
  "profiles": {
    "uhppoted-dll-cli": {
      "commandName": "Project",
      "commandLineArgs": "get-controllers",
      "workingDirectory": "C:\\uhppoted\\uhppoted-dll\\examples\\csharp\\vs-win",
      "environmentVariables": {
        "PATH": "%PATH%;${ProjectDir}"
      }
    }
  }
}
```
and add the DLL folder to the global PATH environment variable, or set the PATH when you run it, e.g.
```
PATH=. && .\bin\Debug\net8.0\uhppoted-dll-cli get-devices
```
(DLL lookup on Windows is insanely complicated - [Dynamic-link library search order](https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order))

Otherwise:

1. I honestly have no idea if the DLL will work with VB.NET - it's only ever been tested with C#. But if you were able to run _GetDevices_ with the debug/test DLL then it should be ok.
2. Are you able to build and run the [_vs-win_](https://github.com/uhppoted/uhppoted-dll/tree/main/examples/csharp/vs-win) example ?
3. Can you perhaps post your project, solutions and launch settings files?

**Reply by threeeye** [reply]
I don't really know how I resolved the error before.

I also converted the uhppoted.cs and lookup.cs from C# to VB.NET.
I tried adding the makefile (and editing it) to my project, when running 'make' from wingw64, the project runs and finds 22 of my 50 controllers.
But when I run the project from VS, I get this error:
`System.DllNotFoundException: 'Unable to load DLL 'uhppoted.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)'`

Here is my code:
```
Public Class Form1
    Dim AQ As uhppoted.Uhppoted
    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Dim controllers() As uhppoted.Controller = Array.Empty(Of uhppoted.Controller)()

        AQ = New uhppoted.Uhppoted("192.168.110.7", "255.255.255.255", "0.0.0.0:60001", 2500, controllers, True)

        Dim c As UInt32() = AQ.GetDevices()

        MsgBox(c.Length)
    End Sub
End Class
```

it crashes on:
```
Public Function GetDevices() As UInteger()
            Dim N As Integer = 0
            Dim count As Integer = N
            Dim slice As UInteger()

            Do
                N += 16
                count = N
                slice = New UInteger(N - 1) {}
                Dim err As String = GetDevices(u, count, slice)

                If err IsNot Nothing AndAlso err <> "" Then
                    Throw New UhppotedException(err)
                End If
            Loop While N < count

            Dim list As UInteger() = New UInteger(count - 1) {}
            Array.Copy(slice, list, list.Length)
            Return list
        End Function
```
in line `Dim err As String = GetDevices(u, count, slice)` (this is in the Uhppoted Class)

EDIT:
Adding launchSettings.json to my project didn't help, it generated an empty App.config (I didn't have that file before)

**Comment by uhppoted** [comment]
re. VB.NET. Huh! Good to know it's possible! Any particular reason you converted to VB rather than just stick with the C#?

But, ok, that helps - assuming your _Makefile_ still looks more or less like the example _Makefile_, when it runs a command it first sets the PATH environment variable to the _lib_ folder and then runs the command. So the actual command it is running is  something like:
```
PATH=..\..\..\lib && .\bin\Debug\net8.0\uhppoted-dll-cli get-devices
```
The  PATH bit allows Windows to find _uhppoted.dll_ in the _lib_ folder.

To get the same behaviour inside VS you need to edit the launch configuration and set the PATH environment variable to `"..\..\..lib"`. Microsoft doesn't seem to have any documentation on how to do that for VB but for C# it's [here](https://learn.microsoft.com/en-us/visualstudio/debugger/project-settings-for-csharp-debug-configurations-dotnetcore?view=vs-2022#launch-profile-net-core-net-5) and I assume it's similar.

So if _uhppoted.dll_ is located in the original _lib_ folder, it would look something like this:
```
...
     "environmentVariables": {
        "PATH": "%PATH%;..\..\..\lib"
      }
...
```
or if you've copied the the DLL to your project folder:
```
...
     "environmentVariables": {
        "PATH": "%PATH%;${ProjectDir}"
      }
...
```

**Reply by threeeye** [reply]
I converted them to VB because I can't have a mixed VB and C# project (as far as I know), and although C# and VB are very similar, I don't get along with C# (I blame the ';' )...
Here is my launchSettings.json:
```
{
  "profiles": {
    "AQ-Test-Core": {
      "commandName": "Project",
      "commandLineArgs": "get-controllers",
      "workingDirectory": "d:\\Projects\\AQ-Test-Core\\AQ-Test-Core",
      "environmentVariables": {
        "PATH": "%PATH%;${ProjectDir}"
      }
    }
  }
}
```
and the App.config that got generated:
```
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
</configuration>
```
And the Makefile:
```
DLL = ..\..\..\lib
CLI = PATH=. && .\bin\Debug\net8.0-windows10.0.22621.0\AQ-Test-Core
CONTROLLER = 0

get-controllers:
	$(CLI) get-controllers

get-controller:
	$(CLI) get-controller --controller $(CONTROLLER)

set-address:
	$(CLI) set-address --controller $(CONTROLLER)

get-status:
	$(CLI) get-status --controller $(CONTROLLER)

get-time:
	$(CLI) get-time --controller $(CONTROLLER)

set-time:
	$(CLI) set-time --controller $(CONTROLLER)

get-listener:
	$(CLI) get-listener --controller $(CONTROLLER)

set-listener:
	$(CLI) set-listener --controller $(CONTROLLER)

get-door-control:
	$(CLI) get-door-control --controller $(CONTROLLER)

set-door-control:
	$(CLI) set-door-control --controller $(CONTROLLER)

open-door:
	$(CLI) open-door --controller $(CONTROLLER)

get-cards:
	$(CLI) get-cards --controller $(CONTROLLER)

get-card:
	$(CLI) get-card --controller $(CONTROLLER)

get-card-by-index:
	$(CLI) get-card-by-index --controller $(CONTROLLER)

put-card:
	$(CLI) put-card --controller $(CONTROLLER)

delete-card:
	$(CLI) delete-card --controller $(CONTROLLER)

delete-all-cards:
	$(CLI) delete-all-cards --controller $(CONTROLLER)

get-event-index:
	$(CLI) get-event-index --controller $(CONTROLLER)

set-event-index:
	$(CLI) set-event-index --controller $(CONTROLLER)
 
get-event:
	$(CLI) get-event --controller $(CONTROLLER)

record-special-events:
	$(CLI) record-special-events --controller $(CONTROLLER) 

get-time-profile:
	$(CLI) get-time-profile --controller $(CONTROLLER)

set-time-profile:
	$(CLI) set-time-profile --controller $(CONTROLLER)

clear-time-profiles:
	$(CLI) clear-time-profiles --controller $(CONTROLLER)

add-task:
	$(CLI) add-task --controller $(CONTROLLER)

refresh-tasklist:
	$(CLI) refresh-tasklist --controller $(CONTROLLER)

clear-tasklist:
	$(CLI) clear-tasklist --controller $(CONTROLLER)

set-pc-control:
	$(CLI) set-pc-control --controller $(CONTROLLER)

set-interlock:
	$(CLI) set-interlock --controller $(CONTROLLER)

activate-keypads:
	$(CLI) activate-keypads --controller $(CONTROLLER)

set-door-passcodes:
	$(CLI) set-door-passcodes --controller $(CONTROLLER)

restore-default-parameters:
	$(CLI) restore-default-parameters --controller $(CONTROLLER)
```

EDIT:
I think I got it!!!
I added a DLLImport to the Form1_Load:
`LoadNativeAssembly(Application.StartupPath, "uhppoted.dll")`
and:
```
    <DllImport("kernel32.dll", CharSet:=CharSet.Auto, SetLastError:=True)>
    Private Shared Function LoadLibrary(libname As String) As IntPtr
    End Function

    Private Sub LoadNativeAssembly(nativeBinaryPath As String, assemblyName As String)
        Dim path = IO.Path.Combine(nativeBinaryPath, assemblyName)
        Dim ptr = LoadLibrary(path)

        If ptr = IntPtr.Zero Then
            Throw New Exception(String.Format("Error loading {0} (ErrorCode: {1})", assemblyName, Marshal.GetLastWin32Error()))
        End If
    End Sub
```

But I still only get 22 out of my 50 controllers

EDIT2:
I'm going on vacation, so I don't know when I'll be able to continue with this project

**Comment by uhppoted** [comment]
Ah! **Great**! I was about to suggest either printing out the `PATH` environment variable just before the DLL gets loaded or else add the folder to the PATH in the system environment variables, but that looks like a better way to do it. Nice!

_re. 22 out of 50 controllers_. That is most likely something to do with the network configuration:
- Are all the controllers on the same LAN segment? If you can post the IP address of one of the '_found_' controllers and the IP address of one the '_not found_' controllers it might provide a clue
- If you run the [uhppote-cli](https://github.com/uhppoted/uhppote-cli) `get-devices` command does it also only return 22 controllers?
```
uhppote-cli --debug get-devices
```

You could also try increasing the timeout from 2500ms to 5000ms - don't think that's the issue but it's worth eliminating:
```
AQ = New uhppoted.Uhppoted("192.168.110.7", "255.255.255.255", "0.0.0.0:60001", 5000, controllers, True)
```

**Reply by threeeye** [reply]
The controllers don't have IP addresses (non of them).
And yes, they are all on the same network, and I tried increasing the ms up to 10000, no change.
And running 'make' on the C# example I get the 3 simulation controllers.

Running the exe:
```
D:\mingw64\uhppoted-dll\examples\csharp\vs-win\bin\Debug\net8.0>uhppoted-dll-cli.exe --debug get-controllers
uhppoted-dll Visual Studio C# Example (MacOS)

  *** ERROR: invalid command (--debug)

  Usage: uhppoted-dll-cli <command>

  Supported commands
    get-controllers
    get-controller
    set-address
    get-status
    get-time
    set-time
    get-listener
    set-listener
    get-door-control
    set-door-control
    open-door
    get-cards
    get-card
    get-card-by-index
    put-card
    delete-card
    delete-all-cards
    get-event-index
    set-event-index
    get-event
    record-special-events
    get-time-profile
    set-time-profile
    clear-time-profiles
    add-task
    refresh-tasklist
    clear-tasklist
    set-pc-control
    set-interlock
    activate-keypads
    set-door-passcodes
    restore-default-parameters


D:\mingw64\uhppoted-dll\examples\csharp\vs-win\bin\Debug\net8.0>uhppoted-dll-cli.exe get-controllers
uhppoted-dll Visual Studio C# Example (MacOS)
>>> get-devices
    N:  16

get-controllers (3)
   201020304
   303986753
   405419896
```

**Comment by uhppoted** [comment]
That suggests it's something to do with the network configuration - possibly the controllers don't have the correct gateway or subnet masks configured.

Can you run the `get-devices` command using the CLI from this project: https://github.com/uhppoted/uhppote-cli. It's the reference implementation and if it also only finds 22 controllers it will more or less confirm the network configuration hypothesis (there are executables in the _Releases_ section so you don't need to build it).

You could also try binding to `0.0.0.0`:
```
AQ = New uhppoted.Uhppoted("0.0.0.0", "255.255.255.255", "0.0.0.0:60001", 2500, controllers, True)
```

The controllers will definitely have IP addresses but they may be being auto-assigned - if the controllers are visible in the manufacturer application the IP addresses should be displayed there somewhere. Or if you have access to the router they may also be in the ARP table.

**Reply by threeeye** [reply]
re. the network configuration, my computer has multiple network cards with multiple IPs on each card (I have my reasons for that).
Binding to `0.0.0.0` resulting in no controllers found (I think it's because of my network setup).
I think that's the reason that the CLI also can't find any controllers.

**Reply by uhppoted** [reply]
You can specify the bind address on the command line:
```
uhppote-cli --bind 192.168.110.7 get-devices
```
That should at least return the 22 controllers - it might be worth doing a sanity check using the other IP addresses. 

Are any of the missing controllers are older than about 10-12 years? The minimum firmware version is 6.56 which is from about 12 years back.

**Reply by threeeye** [reply]
Back from vacation, sorry for the late reply.
I may have a few controllers that have FW lower than 6.56, I'm using the program version 7.83 because the newest version shows the old controllers, but can't monitor them (or do anything).
(I once got an error, something like "!!! 6. (something that I can't remember...) !!!")
Is there a way to update the FW on the controller itself?

**Reply by uhppoted** [reply]
re. vacation. Hope it was good :-) :-).

re. update firmware. Sadly there isn't - the manufacturer has stated categorically that it isn't possible (see the one and only question in the [FAQ](https://github.com/uhppoted/uhppoted/blob/master/documentation/FAQ.md)). I guess it's probably possible for them but the shipping costs and complications and associated risk just make it not viable.

The manufacturer app should (_probably ? somewhere ?_) display the controller firmware version - it might be under the advanced functions menu, otherwise I guess you could try sniffing packets with Wireshark. If you can identify any that are 6.56+ but aren't showing up in _uhppote-cli_ we can try and figure out why.

**Reply by threeeye** [reply]
I'm trying to "monitor" a controller (that I know is with the right FW, I do see it on the list of the 22 that are found)
Reading through the docs, I see a Listener and Event Sub, but I don't see how I set a Handler to "monitor" card swipe and door open/close events and post them...

re. firmware. That's sad, I'll need to replace about 30 controllers :(

**Reply by uhppoted** [reply]
Ouch! Might be worth shopping around - I remember somebody getting _old stock_ (v6.62) controllers off [aliexpress.de](https://www.aliexpress.de) for a very reasonable price (about half the cost of Amazon).

The DLL currently doesn't support event listening because trying to get a callback working in a DLL across half a dozen languages was just too hair tearingly out intimidating at the time. I'll start [looking into it](https://github.com/uhppoted/uhppoted-dll/issues/11) - it's not trivial though so it's going to take some thought and experimentation (i.e. time). There are a couple of approaches you could try in the interim:

- poll the controller with [_GetStatus_](https://github.com/uhppoted/uhppoted-dll/blob/main/doc/csharp.md#uhppotedgetstatus) and if the current event index has changed then retrieve the new events using [_GetEvent_](https://github.com/uhppoted/uhppoted-dll/blob/main/doc/csharp.md#uhppotedgetevent).
- create your own UDP listener and decode the received events
- use something like [uhppoted-app-db](https://github.com/uhppoted/uhppoted-app-db) to store events to a _sqlite3_ database and monitor the database

**Reply by threeeye** [reply]
re. GetStatus, I tried pulling the status with `Dim stat As Status = AQ.GetStatus(123339118)` (this is a real controller that works, I can open and close doors), but my program freezes (not crashing, just freezes).

re. UDP listener, can you give me some pointers/directions?

re. database, I'll use MSSQL for the database

EDIT:
The program freezes also on 'GetDevice':
```
  Dim c As Device = AQ.GetDevice(123339118)
  MsgBox(c.version)
```
I think it's all the 'get'

**Reply by uhppoted** [reply]
123339118 is a very weird serial number? They typically start with 2 (for the two door controllers) or 4 (four door controllers). 

Howsoever .. going to need to narrow it down a bit:

1. Does anything get logged to the console if you enable `debug` in your implementation? In C# you would set it in the constructor:
```
 public Uhppoted(string bind, string broadcast, string listen, int timeout,Controller[] controllers, bool debug) {
   ...
```
2. What timeout value are you using?
3. Can you post the output for _uhppote-cli get device_:
```
uhppote-cli --debug get-device 123339118
```
4. Ditto for the C# example, e.g.:
```
uhppoted-dll-cli get-controller --controller 123339118
```

re. UDP listener. TBH, I haven't a clue how you would do it in VB .. however, as a first approximation:
- start with this _stackoverflow_ answer: https://stackoverflow.com/questions/16160550/visual-basic-udpclient-server-client-model. The second answer is better than the accepted answer - if you're using the first answer then you only need the receive bit.
- set your controller listener address and port with `set-listener`
- swipe a card
- see if you receive anything at all and we can take it from there

**Reply by threeeye** [reply]
re. weird serial number, it's a 1 door controller.
The rest are 4 doors, and their SN starts with 4

re. debug, I think my debug is set to on (aka True): `AQ = New uhppoted.Uhppoted("192.168.110.7", "255.255.255.255", "0.0.0.0:60001", 2500, Array.Empty(Of Controller)(), True)`

re. timeout, I tried with 2500, 5000 and 10000

re. output
From uhppote-cli:
```
D:\uhppote-cli_v0.8.8\windows>uhppote-cli.exe --bind 192.168.110.7 --broadcast 255.255.255.255:60000 --debug get-device 123339118
 ... request
 ...          00000000  17 94 00 00 6e 01 5a 07  00 00 00 00 00 00 00 00  |....n.Z.........|
 ...          00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...

 ... sent 64 bytes to 255.255.255.255:60000

 ... received 64 bytes from 192.168.110.6:60000
 ...          00000000  17 94 00 00 6e 01 5a 07  c0 a8 00 00 ff ff ff 00  |....n.Z.........|
 ...          00000010  00 00 00 00 00 66 07 5a  01 6e 08 98 20 19 05 21  |.....f.Z.n.. ..!|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...
123339118  192.168.0.0     255.255.255.0   0.0.0.0         00:66:07:5a:01:6e v8.98 2019-05-21
```
Output from uhppoted-dll-cli:
```
uhppoted-dll\examples\csharp\vs-win\bin\Debug\net8.0>uhppoted-dll-cli.exe get-controller --controller 123339118
uhppoted-dll Visual Studio C# Example (MacOS)
>>> get-device
    ID: 123339118
```
re. UDP listener, I'll try it

**Reply by uhppoted** [reply]
For some reason your _uhppoted-dll-cli_ is using the `lib/debug/uhppoted.dll` - in which case you should have gotten **exactly** this output:
```
>>> get-device
    ID: 123339118

get-device
  ID              405419896
  address         192.168.1.101
  subnet mask     255.255.255.0
  gateway address 192.168.1.1
  MAC             00:12:23:34:45:56
  version         v8.92
  released        2018-11-05
```

Best guess at this stage is that the timeout is being set to something large because of an alignment mismatch between VB and cgo. Can you post your implementation of `uhppoted.cs`? Specifically the bit where you initialise the `Uhppoted` struct, i.e.:
```
    public Uhppoted(string bind, string broadcast, string listen, int timeout,
                    Controller[] controllers, bool debug) {
        this.u.bind = bind;
        this.u.broadcast = broadcast;
        this.u.listen = listen;
        this.u.timeout = timeout;
        this.u.devices = IntPtr.Zero;
        this.u.debug = debug;
        ...
```

**Reply by threeeye** [reply]
re. uhppoted.cs (which is actually uhppoted.vb):
```
        Public Sub New(bind As String, broadcast As String, listen As String, timeout As Integer, controllers As Controller(), debug As Boolean)
            u.bind = bind
            u.broadcast = broadcast
            u.listen = listen
            u.timeout = timeout
            u.devices = IntPtr.Zero
            u.debug = debug
            Dim N As UInteger = controllers.Length
            Dim list As Udevice() = New Udevice(N - 1) {}

            For ix As Integer = 0 To controllers.Length - 1
                Dim c As Controller = controllers(ix)
                list(ix).ID = c.ID
                list(ix).address = c.address
            Next

            Dim sz As Integer = Marshal.SizeOf(GetType(Udevice))
            Dim p As IntPtr = Marshal.AllocHGlobal(CInt(N) * sz)

            For ix As Integer = 0 To list.Length - 1
                Dim d As Udevice = list(ix)
                Dim q As IntPtr = p + ix * sz
                Marshal.StructureToPtr(d, q, False)
            Next

            Dim devices As New Udevices With {
                .N = N,
                .devices = p
            }
            Dim r As IntPtr = Marshal.AllocHGlobal(Marshal.SizeOf(devices))
            Marshal.StructureToPtr(devices, r, False)
            u.devices = r
        End Sub
```

BTW, when I copy my app to the computer that is currently monitoring all the controllers and I run 'GetDevices' I get the 22 that I saw on my computer, but when I run 'GetStatus' the app just closes (with no error) even when I put it in a Try Catch statement.

Tests for uhppoted-dll-cli on the computer that is currently monitoring all the controllers:

Original DLL, 1st time:
```
C:\Users\AQ\Desktop\net8.0>uhppoted-dll-cli.exe get-controller --controller 123339118
uhppoted-dll Visual Studio C# Example (MacOS)
>>> get-device
    ID: 123339118

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.Runtime.InteropServices.Marshal.FreeCoTaskMem(IntPtr)
   at uhppoted.Uhppoted.GetDevice(UInt32)
   at UhppotedDLLCLI.GetController(uhppoted.Uhppoted, System.String[])
   at UhppotedDLLCLI.Main(System.String[])
```
   
Original DLL, every other time after the 1st time (also with the DLL from uhppoted-dll\lib\debug\):
```
C:\Users\AQ\Desktop\net8.0>uhppoted-dll-cli.exe get-controller --controller 123339118
uhppoted-dll Visual Studio C# Example (MacOS)
>>> get-device
    ID: 123339118
```	

DLL from uhppoted-dll\lib\:
```
C:\Users\AQ\Desktop\net8.0>uhppoted-dll-cli.exe get-controller --controller 123339118
uhppoted-dll Visual Studio C# Example (MacOS)
 ... request
 ...          00000000  17 94 00 00 6e 01 5a 07  00 00 00 00 00 00 00 00  |....n.Z.........|
 ...          00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...

 ... sent 64 bytes to 255.255.255.255:60000

 ... received 64 bytes from 192.168.110.7:60000
 ...          00000000  17 94 00 00 6e 01 5a 07  c0 a8 00 00 ff ff ff 00  |....n.Z.........|
 ...          00000010  00 00 00 00 00 66 07 5a  01 6e 08 98 20 19 05 21  |.....f.Z.n.. ..!|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...
```

 DLL from uhppoted-dll\lib\tests\ (doesn't even see the controller):
```
C:\Users\AQ\Desktop\net8.0>uhppoted-dll-cli.exe get-controller --controller 123339118
uhppoted-dll Visual Studio C# Example (MacOS)
``` 
 
No DLL (fails as expected):
```
 C:\Users\AQ\Desktop\net8.0>uhppoted-dll-cli.exe get-controller --controller 123339118
uhppoted-dll Visual Studio C# Example (MacOS)
  *** ERROR: Unable to load DLL 'uhppoted.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)
```

**Reply by uhppoted** [reply]
Hi,

I've copied most of this to a new [VB.NET discussion](https://github.com/uhppoted/uhppoted/discussions/50) - let's continue there because it's increasingly off topic here.
